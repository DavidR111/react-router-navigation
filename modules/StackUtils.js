/* @flow */
/* eslint no-duplicate-imports: 0 */
/* eslint no-param-reassign: 0 */

import { Children, cloneElement } from 'react'
import { matchPath } from 'react-router'
import type { ContextRouter, Location } from 'react-router'
import type { Route } from './TypeDefinitions'


/**
 * Build stack with React elements
 */
export const build = <Item>(
  children: Array<React$Element<Item>>
): Array<Item & { key: string }> => {
  return Children.toArray(children).reduce((stack, child) => {
    const item = Object.keys(child.props).reduce((props, key) => {
      if (key === 'path') {
        return {
          ...props,
          [key]: child.props[key],
          key: child.props[key],
        }
      } else if (key === 'render' || key === 'component' || key === 'children') {
        return {
          ...props,
          [key]: (ownProps) => {
            return cloneElement(
              child,
              { state: ownProps && ownProps.state },
            )
          },
        }
      }
      return {
        ...props,
        [key]: child.props[key],
      }
    }, {})
    if (!item.key) return stack
    return [...stack, item]
  }, [])
}


/**
 * Test if current stack item should be updated
 */
export const shouldUpdate = (
   currentItem: { path: string, exact?: boolean, strict?: boolean },
   nextItem: { path: string, exact?: boolean, strict?: boolean },
   currentContext: ContextRouter,
   nextContext: ContextRouter,
 ): boolean => {
  const { location: currentLocation } = currentContext
  const { history: { entries, index }, location: nextLocation } = nextContext
  if (entries === undefined || index === undefined) return false
  // Get entries and matchs
  const previousEntry = entries[index - 1]
  const currentEntry = entries[index]
  const nextEntry = entries[index + 1]
  const matchCurrentRoute = matchPath(nextLocation.pathname, currentItem)
  const matchNextRoute = matchPath(nextLocation.pathname, nextItem)
  return (
    // Test if pathames are different
    (currentLocation.pathname !== nextLocation.pathname) &&
    // case 1) basic pathname
    ((currentItem.path !== nextItem.path) ||
    // case 2) pathname with query params
    // ex: with same path article/:id,
    //     pathname article/2 !== article/3
    (matchCurrentRoute !== null && matchNextRoute !== null &&
     Object.keys(matchCurrentRoute.params).length !== 0 &&
     Object.keys(matchNextRoute.params).length !== 0 &&
     ((previousEntry && currentEntry.pathname !== previousEntry.pathname) ||
      (nextEntry && currentEntry.pathname !== nextEntry.pathname))
    ))
  )
}


/**
 * Get stack item from a specific route
 * $FlowFixMe
 */
export const get = <Item>(items: Array<Item>, route: Route): ?Item => ({
  ...route,
  ...items.find((item) => {
    return (
      item && item.key &&
      route.routeName === item.key
    )
  }),
})


/**
 * Generate unique key
 */
export const createKey = (route: Route): string => {
  return `${route.key}@@${Math.random().toString(10)}`
}


/**
 * Get current route from a specific history location
 */
export const getRoute = (stack: Array<Object>, location: Location): ?Route => {
  const { pathname } = location
  const item = stack.find((stackItem) => {
    return matchPath(pathname, stackItem)
  })
  if (!item || !item.key) return null
  return {
    key: createKey(item),
    routeName: item.key,
    match: matchPath(pathname, item),
  }
}


/**
 * Render a subview with props
 */
export const renderSubView = (
  render: Function,
  ownProps?: any = {},
) => (props: any): ?React$Element<any> => {
  const allProps = { ...ownProps, ...props }
  const { cards, tabs, scene, route, navigationState: { routes, index } } = allProps
  const item = get(
    cards || tabs,
    (scene && scene.route) || route || routes[index],
  )
  if (!item) return null
  return render({ ...allProps, ...item })
}
