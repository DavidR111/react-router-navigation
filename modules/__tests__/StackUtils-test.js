import React, { createElement, cloneElement } from 'react'
import { Text } from 'react-native'
import { Router, Route } from 'react-router'
import createHistory from 'history/createMemoryHistory'
import renderer from 'react-test-renderer'
import StackUtils from './../StackUtils'

const { build, shouldUpdate, get, getRoute, createKey } = StackUtils

describe('StackUtils.build() util', () => {
  it('build() works correctly', () => {
    const children = [
      <Route
        path="hey"
        render={() => null}
      />,
      <Route
        path="hello"
        component={() => null}
        title="Hello"
      />
    ]
    const results = [{
      path: 'hey',
      key: 'hey',
      render: () => nulll,
    }, {
      path: 'hello',
      key: 'hello',
      component: () => null,
      title: 'Hello',
    }]
    expect(JSON.stringify(build(children))).toEqual(JSON.stringify(results))
  })

  it('build() creates correctly render method', () => {
    const children = [
      <Route
        path="/"
        render={({ path }) => <Text>{path}</Text>}
      />
    ]
    const stack = build(children)
    const history = createHistory()
    const component = renderer.create(
      <Router history={history}>
        {stack[0].render()}
      </Router>
    )
    const tree = component.toJSON()
    expect(tree).toMatchSnapshot()
  })
})

describe('StackUtils.shouldUpdate() util', () => {
  it('shouldUpdate() returns true when paths are different', () => {
    const currentCard = { path: '/foo' }
    const nextCard = { path: '/bar' }
    const currentRouterHistory = createHistory({
      initialEntries: ['/foo'],
    })
    const nextRouterHistory = createHistory({
      initialIndex: 1,
      initialEntries: ['/foo', '/bar'],
    })
    const input = shouldUpdate(currentCard, nextCard, currentRouterHistory, nextRouterHistory)
    expect(input).toBeTruthy()
  })
  it('shouldUpdate() returns false when paths are equal', () => {
    const currentCard = { path: '/foo' }
    const nextCard = { path: '/foo' }
    const currentRouterHistory = createHistory({
      initialEntries: ['/foo'],
    })
    const nextRouterHistory = createHistory({
      initialIndex: 0,
      initialEntries: ['/foo'],
    })
    const input = shouldUpdate(currentCard, nextCard, currentRouterHistory, nextRouterHistory)
    expect(input).toBeFalsy()
  })
  it('shouldUpdate() returns true when paths (with params) are equal and pathnames are different (1)', () => {
    const currentCard = { path: '/article/:id' }
    const nextCard = { path: '/article/:id' }
    const currentRouterHistory = createHistory({
      initialIndex: 1,
      initialEntries: ['/article/1', '/article/2'],
    })
    const nextRouterHistory = createHistory({
      initialEntries: ['/article/1', '/article/2'],
    })
    const input = shouldUpdate(currentCard, nextCard, currentRouterHistory, nextRouterHistory)
    expect(input).toBeTruthy()
  })
  it('shouldUpdate() returns true when paths (with params) are equal and pathnames are different (2)', () => {
    const currentCard = { path: '/article/:id' }
    const nextCard = { path: '/article/:id' }
    const currentRouterHistory = createHistory({
      initialEntries: ['/article1'],
    })
    const nextRouterHistory = createHistory({
      initialIndex: 1,
      initialEntries: ['/article/1', '/article/2'],
    })
    const input = shouldUpdate(currentCard, nextCard, currentRouterHistory, nextRouterHistory)
    expect(input).toBeTruthy()
  })
  it('shouldUpdate() returns false when paths (without params) are equal and pathname are different', () => {
    const currentCard = { path: '/app' }
    const nextCard = { path: '/app' }
    const currentRouterHistory = createHistory({
      initialEntries: ['/app'],
    })
    const nextRouterHistory = createHistory({
      initialIndex: 1,
      initialEntries: ['/app', '/app/article'],
    })
    const input = shouldUpdate(currentCard, nextCard, currentRouterHistory, nextRouterHistory)
    expect(input).toBeFalsy()
  })
})

describe('StackUtils.get() util', () => {
  it('get() works correctly', () => {
    const route = { key: '/index@@h9208990', routeName: '/index' }
    const currentCard = {
      key: '/index',
      title: 'Index',
    }
    const cards = [{
      key: '/',
      title: 'Root',
    }, currentCard]
    expect(get(cards, route)).toEqual(currentCard)
  })
})

describe('StackUtils.getRoute() util', () => {
  it('getRoute() works correctly', () => {
    const stack = [{ key: '/foo', path: '/foo' }, { key: '/bar', path: '/bar' }]
    const history = createHistory({ initialEntries: ['/bar'] })
    const currentRoute = { key: '/bar', routeName: '/bar' }
    expect(getRoute(stack, history.location)).toEqual(currentRoute)
  })
})

describe('StackUtils.createKey() util', () => {
  it('createKey() works correctly', () => {
    const key = createKey({ key: '/foo' })
    expect(key.slice(0, 4)).toEqual('/foo')
  })
})
