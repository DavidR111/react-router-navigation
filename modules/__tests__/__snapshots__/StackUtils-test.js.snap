exports[`StackUtils.build() util build() creates correctly render method 1`] = `
<Text
  accessible={true}
  allowFontScaling={true}
  ellipsizeMode="tail">
  /
</Text>
`;
